src/main/sensors/compass.c

```diff
@@ -388,31 +398,233 @@ uint32_t compassUpdate(timeUs_t currentTimeUs)
     }
 
     flightDynamicsTrims_t *magZero = &compassConfigMutable()->magZero;
-    if (magInit) {              // we apply offset only once mag calibration is done
-        mag.magADC[X] -= magZero->raw[X];
-        mag.magADC[Y] -= magZero->raw[Y];
-        mag.magADC[Z] -= magZero->raw[Z];
-    }
 
     if (tCal != 0) {
         if ((currentTimeUs - tCal) < 30000000) {    // 30s: you have 30s to turn the multi in all directions
             LED0_TOGGLE;
-            for (int axis = 0; axis < 3; axis++) {
-                if (mag.magADC[axis] < magZeroTempMin.raw[axis])
-                    magZeroTempMin.raw[axis] = mag.magADC[axis];
-                if (mag.magADC[axis] > magZeroTempMax.raw[axis])
-                    magZeroTempMax.raw[axis] = mag.magADC[axis];
-            }
+
+            const float recipdTimeSec = 1.0f / (dTimeUs * 1.0e-6f);
+            const float dmag[3] =  {(mag.magADC[X] - magPrevious.magADC[X]) * recipdTimeSec,
+                                    (mag.magADC[Y] - magPrevious.magADC[Y]) * recipdTimeSec,
+                                    (mag.magADC[Z] - magPrevious.magADC[Z]) * recipdTimeSec};
+            const float gyroADCfRadians[3] =  {DEGREES_TO_RADIANS(gyro.gyroADCf[X]),
+                                               DEGREES_TO_RADIANS(gyro.gyroADCf[Y]),
+                                               DEGREES_TO_RADIANS(gyro.gyroADCf[Z])};
+            compassBiasEstimatorApply(&compassBiasEstimator, mag.magADC, dmag, gyroADCfRadians);
+
         } else {
             tCal = 0;
             for (int axis = 0; axis < 3; axis++) {
-                magZero->raw[axis] = (magZeroTempMin.raw[axis] + magZeroTempMax.raw[axis]) / 2; // Calculate offsets
+                magZero->raw[axis] = lrintf(compassBiasEstimator.b[axis]);
             }
 
             saveConfigAndNotify();
         }
     }
 
+    magPrevious = mag;
+
+    for (int axis = 0; axis < 3; axis++) {
+        DEBUG_SET(DEBUG_MAG_BIAS, axis, lrintf(compassBiasEstimator.b[axis]));
+    }
+    // map adaptive forgetting factor lambda from (lambda_min, 1.0f) -> (0, 2000)
+    const float mapLambdaGain = 1.0f / ( 1.0f - compassBiasEstimator.lambda_min) * 2.0e3f;
+    DEBUG_SET(DEBUG_MAG_BIAS, 3, lrintf((compassBiasEstimator.lambda - compassBiasEstimator.lambda_min + 1.0e-6f) * mapLambdaGain));
```

Comment:
To avoid division by 0 (in case anyone ever uses lambda_min = 1.0f). This adjustment is intended solely for debugging purposes, and adding 1.0e-6f should not result in any noticeable differences.

Reply:
I seems that division was refactored to (or is on) previous line.

Reply:
Thx, yes, I placed 1.0e-6f a line above were it should have been from the beginning. Just to guarantee that a division by zero is avoided. If you suggest an alternative way I'm more than happy to change it accordingly.